src/messaging/ExchangeContext.cpp

```diff
@@ -111,48 +115,69 @@ CHIP_ERROR ExchangeContext::SendMessage(Protocols::Id protocolId, uint8_t msgTyp
     bool isUDPTransport                = peerAddress && peerAddress->GetTransportType() == Transport::Type::kUdp;
     bool reliableTransmissionRequested = isUDPTransport && !sendFlags.Has(SendMessageFlags::kNoAutoRequestAck);
 
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
     // If a response message is expected...
     if (sendFlags.Has(SendMessageFlags::kExpectResponse))
     {
         // Only one 'response expected' message can be outstanding at a time.
-        if (IsResponseExpected())
-        {
-            // TODO: add a test for this case.
-            return CHIP_ERROR_INCORRECT_STATE;
-        }
+        // TODO: add a test for this case.
+        VerifyOrExit(!IsResponseExpected(), err = CHIP_ERROR_INCORRECT_STATE);
 
         SetResponseExpected(true);
 
         // Arm the response timer if a timeout has been specified.
         if (mResponseTimeout > System::Clock::Zero)
         {
-            CHIP_ERROR err = StartResponseTimer();
-            if (err != CHIP_NO_ERROR)
-            {
-                SetResponseExpected(false);
-                return err;
-            }
+            err = StartResponseTimer();
+            VerifyOrExit(err == CHIP_NO_ERROR, SetResponseExpected(false));
         }
     }
 
+    err = mDispatch->SendMessage(mSecureSession.Value(), mExchangeId, IsInitiator(), GetReliableMessageContext(),
+                                 reliableTransmissionRequested, protocolId, msgType, std::move(msgBuf));
+    if (err != CHIP_NO_ERROR && IsResponseExpected())
+    {
+        CancelResponseTimer();
+        SetResponseExpected(false);
+    }
+
+    // Standalone acks are not application-level message sends.
+    if (err == CHIP_NO_ERROR && !isStandaloneAck)
     {
-        // Create a new scope for `err`, to avoid shadowing warning previous `err`.
-        CHIP_ERROR err = mDispatch->SendMessage(mSession.Value(), mExchangeId, IsInitiator(), GetReliableMessageContext(),
-                                                reliableTransmissionRequested, protocolId, msgType, std::move(msgBuf));
-        if (err != CHIP_NO_ERROR && IsResponseExpected())
+        MessageHandled();
+    }
+
+exit:
```

Comment:
The community doesn't like exit blocks any more :). Maybe `MessageHandled` would be a good place to put the logic below in? It automatically closes not needed exchanges, so you could probably just check `mExchangeMgr->GetNumActiveExchanges() > 0`. Also, I would recommend extracting the code below to a separate method like `UpdateSEDPollingMode` for better readability.

Reply:
If it comes to the code extraction - ok. But why is exit block wrong here? I believe it's the most appropriate solution for handling early error returns that result in SetResponseExpected(false);. Putting logic below `MessageHandled` will not consider error cases.

Reply:
1. `MessageHandled` is not only called for sent messages, but also for received ones, which may close some exchanges.
2. In general, there's effort being put to refactor all the existing code from using "exit" blocks to RAII. For example, see `auto deferred = MakeDefer([&]()` usage. Therefore, I think that switching back to the exit block will cause that at some point somebody will come and rewrite it to RAII approach, again, so maybe we could avoid some unnecessary noise.
3. Also, should we care about early returns? If we exit early and don't add any exchange, should SED mode be modified?

Reply:
Fixed, please verify if appropriate